### DISCUSSION TITLE
How to read response headers, ToHeaders unintuitive

### DISCUSSION THREAD

**Discussion by nostalgictesla** [earthboundkid/requests#46] [Support] [answered]
Is there a simple way to get the HTTP headers from a response?

Intuitively I tried to chain a ToHeaders into my request builder chain and had some trouble debugging why I was getting no response body on my ToJSON call immediately after, I realized the mistake after reading the doc line for ToHeaders and seeing that it issues a HEAD request and consumes the body. This seems very unintuitive since I can't find a simple way to get the response headers for a request and the ToHeaders method does something significantly different from the other ToX methods.

What I was hoping to achieve with ToHeaders:

```Go
	headers := make(map[string][]string, 0)
	var s string
	err := requests.
		URL("https://httpbin.org/base64/SFRUUEJJTiBpcyBhd2Vzb21l").
		ToHeaders(headers).
		ToString(&s).
		Fetch(context.Background())
	if err != nil {
		panic(err)
	}
	for _, v := range headers {
		fmt.Printf("%+v\n", v)
	}
	fmt.Println(s)
```

**Comment by earthboundkid** [comment, marked as answer]
There are two “ToHeaders”. One is a method on Builder. As you found, it issues a HEAD request, so there’s no body to read. The other is a function (requests.ToHeaders). It only copies the headers and doesn’t mess with the HTTP verb. In general, the methods vs functions follow this pattern where the functions are more limited. For example, BodyJSON the method sets a content type of application/json, but the function does not. This should probably be better documented somewhere. In this case, use .AddValidator(requests.ToHeaders) and you can just copy the headers without touching the HTTP verb.

**Comment by nostalgictesla** [comment]
Thanks that makes complete sense. It would be useful to have that documented, perhaps as an example in the README that provides multiple method chains for getting data out of the response. or as a small entry in the FAQ in the bottom.

**Reply by jasonlotito** [reply]
Sample here for those that might be confused. Came across needing this today, and hey, this thread was here! Thanks @nostalgictesla for asking, and yes, would be good to add this to the README.
```
   h := make(map[string][]string)
   // ...
   err := requests.URL("http://www.example.com").
        BodyJSON(req).
        ToJSON(&res).
        ContentType("application/json").
        AddValidator(requests.ToHeaders(h)).
        Fetch(context.Background())
    // ...
    log.Printf("%v", h)

```

**Comment by earthboundkid** [comment]
I thought about this some over the weekend, and the `ToX` methods/functions are handlers, meaning they are meant to consume the response body. Technically, handlers and validators have the same signature: they look at a response and potentially return an error. But you can only read a body once unless you swap it out with a buffer, so there's generally only one handler that looks at the body. 

So, with that in mind, it might make sense to rename `requests.ToHeaders` to `requests.CheckHeaders` and add a `Builder.CheckHeaders` method. `Builder.ToHeaders` can stay, but it would be explicitly meant for HEAD requests where all you do is get the headers and terminate the connection.

What do you all think?

**Reply by nostalgictesla** [reply]
This makes sense. Having the method directly on the builder chain is very intuitive since getting the headers doesn't feel like a validation, just "fetching" or "getting" a routine part of the request for processing. Regarding the name maybe `Builder.GetHeaders` or `Builder.CopyHeaders` would be more descriptive since it isn't checking anything, just copying into a user accessible location. I'd go for `Builder.CopyHeaders`.
